feat(ODC-451): add function that checks if repo is reachable and branch exists

MatousJobanek opened the pull request:
the `IsReachableWithBranch` function checks whether the repo is reachable or not. If it is (which means that the URL is correct and repo is public), then it checks if the provided branch exists
it is needed for this: https://github.com/redhat-developer/devconsole-operator/pull/118

MatousJobanek commented on pkg/git/repository/connectivity.go:
```diff
@@ -0,0 +1,68 @@
+	url := fmt.Sprintf("https://%s%s/info/refs?service=git-upload-pack", endpoint.Host, path)
+	resp, err := client.Get(url)
+	if err != nil || endpoint.Host == "" || endpoint.Path == "" {
+		return false, fmt.Errorf(unableReachRepoError)
```
yes, or more likely fix the wrong if statement - fixed 9169c26

tinakurian commented on pkg/git/repository/connectivity.go:
```diff
@@ -0,0 +1,71 @@
+
+	if resp.StatusCode != http.StatusOK {
+		return false, fmt.Errorf(unableReachRepoError)
+	} else {
```
can we please remove the else statement here? it doesn't look like it's necessary :-)

MatousJobanek replied:
yeah, good point - it's a zombie after refactoring of the final code 227e92b

tinakurian commented on pkg/git/repository/connectivity.go:
```diff
@@ -0,0 +1,71 @@
+		if err != nil {
+			log.Error(err, "error while parsing regexp of the given branch")
+			return false, fmt.Errorf(unableParseBranchError)
+		} else {
```
can we please remove the else clause here? we can probably just have 
```
if compile.Find(body) != nil { 
     return true, nil
}

return false, fmt.Errorf(unableFindBranchError)
```

MatousJobanek replied:
yup, done 227e92b

tinakurian commented on pkg/git/repository/connectivity.go:
```diff
@@ -0,0 +1,71 @@
+	unableParseBranchError = "unable to parse the branch"
+)
+
+func IsReachableWithBranch(log *log.GitSourceLogger, branch string, endpoint *gittransport.Endpoint) (bool, error) {
```
could we please add a func comment here describing what the function does?

MatousJobanek replied:
added here 919a1a0

tinakurian commented on pkg/git/repository/connectivity.go:
```diff
@@ -0,0 +1,71 @@
+	resp, err := client.Get(url)
+	if err != nil {
+		return false, fmt.Errorf(unableReachRepoError)
+	} else {
```
can we remove this else clause and just have 

`return validateBranch(log, branch, resp)`

MatousJobanek replied:
done 227e92b

MatousJobanek commented:
@tinakurian I changed it a bit more as I incorporated also the changes from [devconsole-api](https://github.com/redhat-developer/devconsole-api/commit/36791a68b58dd10ed2ead21974bb9101b4ebe51e) repo and aligned with the new validations - see: https://github.com/redhat-developer/devconsole-git/pull/26. All your comments should be addressed
